### PR TITLE
Number

### LINKED ISSUE #462: Improve Number class numeric value handling

### PR DESCRIPTION
**Pull request by sagikazarmark**
Did a massive refactor on Number usage. For now, calculators do the conversion internally where necessary, but in the future we could just require a number in the calculator itself, but this is up to a future debate.

Fixes #462

### CODE REVIEW

**Comment by sagikazarmark** [comment]
Also, this PR fixes a number of local aware decimal separator issues with bcmath and gmp.

**Comment by frederikbosch** [comment]
Great idea!

File: src/Number.php [hunk 1]
```diff
@@ -38,6 +38,56 @@ public function __construct($integerPart, $fractionalPart = '')
...
+    }
+
+    /**
+     * @param float|int|string $number
+     *
+     * @return self
+     */
+    public static function fromNumber($number)
+    {
+        if (is_float($number)) {
```
Comment (frederikbosch) [inline src/Number.php]: I would not use `elseif` here. But rather only a new if statement, or maybe, even a `switch` statement.

```php
switch (gettype($number)) {
  case 'integer':
   ...
}
```

The above saves a one or more function calls.

Comment (sagikazarmark) [reply src/Number.php]: Using gettype for type checking is generally not advised, but using separate ifs is probably cleaner.

**Comment by sagikazarmark** [comment]
(Note in advance: this PR became a bit bigger, so I don't want to squash commits, there are a few meaningful ones)

**Comment by sagikazarmark** [comment]
Thanks for the review!

**Pull request merged**